### PR TITLE
Pod exec enhancements

### LINKED ISSUE #322: cannot retrieve the command exit code after a pod exec
Hi,

I'm doing a pod exec using `client.CoreV1Api(WsApiClient()).connect_get_namespaced_pod_exec()`, very similar to what is done in the `pod_exec.py` example. I would like to get the command exit code but it seems this is not possible yet.

I have a draft patch that updates `kubernetes_asyncio/stream/ws_client.py` to store the content of `ERROR_CHANNEL` into a new field of the `WsApiClient`. This field can later be retrieved by the user.

```
--- a/kubernetes_asyncio/stream/ws_client.py
+++ b/kubernetes_asyncio/stream/ws_client.py
@@ -53,6 +53,7 @@ class WsApiClient(ApiClient):
                  cookie=None, pool_threads=1, heartbeat=None):
         super().__init__(configuration, header_name, header_value, cookie, pool_threads)
         self.heartbeat = heartbeat
+        self.error_data = None
 
     async def request(self, method, url, query_params=None, headers=None,
                       post_params=None, body=None, _preload_content=True,
@@ -82,6 +83,7 @@ class WsApiClient(ApiClient):
         if _preload_content:
 
             resp_all = ''
+            error_data = ''
             async with self.rest_client.pool_manager.ws_connect(url, headers=headers, heartbeat=self.heartbeat) as ws:
                 async for msg in ws:
                     msg = msg.data.decode('utf-8')
@@ -91,6 +93,11 @@ class WsApiClient(ApiClient):
                         if data:
                             if channel in [STDOUT_CHANNEL, STDERR_CHANNEL]:
                                 resp_all += data
+                            elif channel == ERROR_CHANNEL:
+                                error_data += data
+
+            if error_data:
+                self.error_data = error_data
 
             return WsResponse(200, resp_all.encode('utf-8'))
 ```

On success, `error_data` looks like this: `{"metadata":{},"status":"Success"}`.
On error, it looks like this: `{"metadata":{},"status":"Failure","message":"command terminated with non-zero exit code: error executing command [/bin/false], exit code 1","reason":"NonZeroExitCode","details":{"causes":[{"reason":"ExitCode","message":"1"}]}}`

Is it something that would make sense? If yes, I can submit a PR, otherwise do you have an alternative way of retrieving the exit code?

Thanks!

### PR DESCRIPTION
**Pull request by olivier-matz-6wind**
This draft pull request provides some enhancements to the pod_exec example and ws API.
The last commit is an API break, so you may not want it.

Comments welcome, of course!

Fixes #322

### CODE REVIEW

File: kubernetes_asyncio/stream/ws_client.py [hunk 1]
```diff
@@ -91,9 +107,14 @@ async def request(self, method, url, query_params=None, headers=None,
...
+
+            if error_data:
+                self.parse_error_data(error_data)
 
             return WsResponse(200, resp_all.encode('utf-8'))
 
         else:
 
-            return await self.rest_client.pool_manager.ws_connect(url, headers=headers, heartbeat=self.heartbeat)
+            return self.rest_client.pool_manager.ws_connect(url, headers=headers, heartbeat=self.heartbeat)
```
Comment (tomplus) [inline kubernetes_asyncio/stream/ws_client.py:120]: Yes, it's a breaking change. I agree that a context manager would be helpful, so we can do it but it'll be released in a month or so (version 31.x, K8s v1.31 is not released yet).

File: examples/pod_exec.py [hunk 2]
```diff
@@ -1,52 +1,148 @@
...
+            command=exec_command,
+            stderr=True,
+            stdin=False,
+            stdout=True,
+            tty=False,
+        )
+        print(f"Response: {ret}")
+        print(f"Exit code: {ws_api.returncode}")
 
+    # Execute a command interactively. If _preload_content=False is passed to
```
Comment (tomplus) [inline examples/pod_exec.py:96]: Thanks, this example will close this issue too: https://github.com/tomplus/kubernetes_asyncio/issues/290

**Review by tomplus — CHANGES_REQUESTED** [review]
Excellent job! I really appreciate that. Thank you.

File: kubernetes_asyncio/stream/ws_client.py [hunk 3]
```diff
@@ -53,6 +55,18 @@ def __init__(self, configuration=None, header_name=None, header_value=None,
                  cookie=None, pool_threads=1, heartbeat=None):
         super().__init__(configuration, header_name, header_value, cookie, pool_threads)
         self.heartbeat = heartbeat
+        self.returncode = None
```
[the remark this answers is not in the document]
Comment (olivier-matz-6wind) [reply kubernetes_asyncio/stream/ws_client.py:58]: Indeed. I felt that it was not ideal but I had not thought that two coroutines could be executed in parallel.

What about storing the returncode in the WsResponse object instead?

File: kubernetes_asyncio/stream/ws_client.py [hunk 4]
```diff
@@ -53,6 +55,18 @@ def __init__(self, configuration=None, header_name=None, header_value=None,
...
+
+    def parse_error_data(self, error_data):
+        """
+        Parse data received on ERROR_CHANNEL, and set self.returncode accordingly.
+        """
+        error_data_json = json.loads(error_data)
```
[the remark this answers is not in the document]
Comment (olivier-matz-6wind) [reply kubernetes_asyncio/stream/ws_client.py:64]: Yes, I'll change this in next version

File: kubernetes_asyncio/stream/ws_client.py (hunk 1 above)
Comment (olivier-matz-6wind) [reply kubernetes_asyncio/stream/ws_client.py:120]: Sure, do what you feel the most relevant for this patch. On my side I can easily adapt my code.

**Comment by olivier-matz-6wind** [comment]
> Excellent job! I really appreciate that. Thank you.

Thank you for the feedback!

File: kubernetes_asyncio/stream/ws_client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomplus) [reply kubernetes_asyncio/stream/ws_client.py:58]: It won't help, the WsResponse is not returned to a caller.

[the remark this answers is not in the document]
Comment (olivier-matz-6wind) [reply kubernetes_asyncio/stream/ws_client.py:58]: oh right...
I'll update the PR with all these changes then. Thanks

**Review by tomplus — APPROVED** [review]
Thanks for your contribution.
I'm merging it now and it will be released with the next "big" version.

**Pull request merged**